# Лабораторная работа 3.1

## Счастливый билетик

![](http://f2.mylove.ru/UNiHT7dRJ4.jpg)

Билет считается счастливым, если выполнено следующее условие: сумма первых трёх цифр номера равна сумме последних трёх цифр.

**Задание:**

1) Определите число `ticket_number` — шестизначный номер билета;

2) Напишите код, который по шестизначному номеру `ticket_number` билетика проверяет, является ли он счастливым;

3) Если номер счастливый, выведите строку `Yes`, иначе — `No`.

---

**Пример 1:**

*Input:* 123456

*Output:* No


**Пример 2:**

*Input:* 123042

*Output:* Yes

In [1]:
ticket_number = int(input("Enter a number:"))
first3 = ticket_number // 1000
last3 = ticket_number % 1000
if (first3 // 100 + first3 // 10 % 10 + first3 % 10 == last3 // 100 + last3 // 10 % 10 + last3 % 10):
  print("Yes")
else:
  print("No")

Enter a number:567345
No


## Пароль

![](http://forum.ixbt.com/post.cgi?id=attach:77:14138:1341:1)

Пусть пароль может содержать только латинские буквы, знаки препинания и цифры.

Пароль считается надёжным, если удовлетворяет следующим условиям:
- содержит буквы в разных регистрах;
- содержит цифры;
- содержит не менее `4` уникальных символов;
- не содержит ваше имя латинницей, записанное буквами любых регистров (anna, iVan, ...).

Иначе пароль считается слабым.

**Задание:**

1) Определите строку `password` — придуманный вами пароль;

2) Напишите код, который по паролю `password` проверяет, является ли он надёжным;

3) Если пароль надёжный, выведите строку `strong`, иначе — `weak`.


---

Пусть имя пользоватея -- Андрей.

**Пример 1:**

*Input:* Aandrei123

*Output:* weak


**Пример 2:**

*Input:* an12dRei

*Output:* strong

In [85]:
password = input("Enter a password: ")
name = input("Enter a name: ")
if (password == password.upper() or password == password.lower() or password.isalpha() 
   or len(set(password)) < 4 or name.lower() in password.lower()):
  print("weak")
else:
  print("strong")

Enter a password: yhdvgYU74
Enter a name: yana
strong


## Числа Фибоначчи
Как известно, [числа Фибоначчи](https://ru.wikipedia.org/wiki/%D0%A7%D0%B8%D1%81%D0%BB%D0%B0_%D0%A4%D0%B8%D0%B1%D0%BE%D0%BD%D0%B0%D1%87%D1%87%D0%B8) — это последовательность чисел, каждое из которых равно сумме двух предыдущих (первые два числа равны $1$):
$$
    1, 1, 2, 3, 5, 8, 13, \ldots
$$

**Задание:**

1) Определите число `amount` — количество чисел Фибоначчи, которые надо вывести;

2) Напишите код, который выводит первые `amount` чисел Фибоначчи.

---

**Пример 1:**

*Input:* 3

*Output:* 1 1 2


**Пример 2:**

*Input:* 10

*Output:* 1 1 2 3 5 8 13 21 34 55


In [6]:
amount = int(input("Enter a numbe: "))
a, b = 0, 1
print(b)
for i in range(amount):
    sum = a + b
    a = b
    b = sum
    print(b)

Enter a numbe: 6
1
1
2
3
5
8
13


## Время исследований

![](https://2.bp.blogspot.com/-VOT5x12UqPY/WSh53tQ94HI/AAAAAAAAAMY/-nlcz28_Fj4UzJjbn4bBINKCnzwag1W4QCLcB/s1600/my-pic-widget.jpg)

На сайте https://www.kaggle.com/ выберите любой набор данных в формате CSV и проведите для него маленькое исследование: загрузите данные из набора с использованием стандартного модуля `csv`, посмотрите средние значения и стандартные отклонения двух выбранных числовых атрибутов, найдите [методом наименьших квадратов](https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D1%82%D0%BE%D0%B4_%D0%BD%D0%B0%D0%B8%D0%BC%D0%B5%D0%BD%D1%8C%D1%88%D0%B8%D1%85_%D0%BA%D0%B2%D0%B0%D0%B4%D1%80%D0%B0%D1%82%D0%BE%D0%B2) уравнение линейной зависимости, связывающей один числовой атрибут с другим. Для оценки заданной зависимости найдите [коэффициент парной корреляции](https://ru.wikipedia.org/wiki/%D0%9A%D0%BE%D1%80%D1%80%D0%B5%D0%BB%D1%8F%D1%86%D0%B8%D1%8F), сделайте соответствующие выводы.

**Результаты надо обязательно прокомментировать и пояснить!**

---

**Пример 1:**

Пусть таблица `bikes.csv` содержит данные по арендам велосипедов за $2$ года:
- `datetime`: дата и время аренды
- `season`: время года
- `temp`: температура воздуха по Цельсию
- `windspeed`: скорость ветра 
- `registered`: число аренд

Одно из направлений исследования могло бы заключаться в проверке зависимости суммарного числа аренд от температуры воздуха.

Импорт библиотек:

Чтение файла с данными и вывод столбцов:

In [66]:
import csv
from math import sqrt


with open('avocado.csv', 'r', newline='') as csvf:
    data = csv.reader(csvf, delimiter=',')
    total_vol = []
    total_bags = []
    for row in data:
        if row[3] == "Total_Volume":
            continue
        total_vol.append(float(row[3]))
        total_bags.append(float(row[7]))

Подсчитаем среднее значение в 2-х столбцах

In [67]:
vol = sum(total_vol) / len(total_vol)
bag = sum(total_bags) / len(total_bags)
print(f"Среднее значение Total_Volume {vol}")
print(f"Среднее значение Total_Bags : {bag}")

Среднее значение Total_Volume 850644.0130089332
Среднее значение Total_Bags : 239639.20205983953


Подсчитаем стандартные отклонения двух выбранных числовых атрибутов:

In [74]:
v1 = sum((el-vol)**2 for el in total_vol) / len(total_vol)
st_v = sqrt(v1)
v2 = sum((elem-bag)**2 for elem in total_bags) / len(total_bags)
st_b = sqrt(v2)
print(sum((el-vol)**2 for el in total_vol))
print(len(total_vol))
print(f"Стандартное отклонение Total_Volume: {st_v}")
print(f"Стандартное отклонение Total_Bags: {st_b}")

2.176434493218307e+17
18249
Стандартное отклонение Total_Volume: 3453450.731237387
Стандартное отклонение Total_Bags: 986215.3770258684


Проверила с помощью ддугой библиотеки

In [75]:
import numpy as np 
import pandas as pd 
data = pd.read_csv('avocado.csv')
data.std()

C:\Users\ynakh\AppData\Local\Temp\ipykernel_11696\1959062251.py:4: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data.std()


Unnamed: 0      1.548104e+01
AveragePrice    4.026766e-01
Total_Volume    3.453545e+06
4046            1.264989e+06
4225            1.204120e+06
4770            1.074641e+05
Total_Bags      9.862424e+05
Small_Bags      7.461785e+05
Large_Bags      2.439660e+05
XLarge_Bags     1.769289e+04
year            9.399385e-01
dtype: float64

Составим уравнение линейной зависимости

In [61]:
sum_ab = 0
sum_square = 0

for i, el in enumerate(total_vol):
    sum_ab += el * total_bags[i]
    sum_square += el**2

size = len(total_vol)
k_lin = (size * sum_ab - sum(total_vol) * sum(total_bags))/(size * sum_square - sum(total_vol)**2)
b_lin = vol - bag * k_lin
func_val = []

for el in total_vol:
    func_val.append(k_lin * el + b_lin)

print(f"Уравнение линейной зависимости: y = {k_lin}x + {b_lin}")

Уравнение линейной зависимости: y = 0.2750211065408285x + 784738.1744878748


Найдем коэффициент парной корреляции:

In [63]:
cor_chisl = 0
for i, el in enumerate(total_vol):
    cor_chisl += (el - vol)*(total_bags[i] - bag)
sqr_diff_vol = sum((el-vol)**2 for el in total_vol)
sqr_diff_bag = sum((el-bag)**2 for el in total_bags)
r_xy = cor_chisl / sqrt(sqr_diff_vol * sqr_diff_bag)
print(f"Коэффициент парной корреляции: {r_xy}")

Коэффициент парной корреляции: 0.9630470824267295


Коэффициент парной корреляции ~ 0.96, что может говорить о неявной зависимости